community/google-summer-of-code/reports/2017/lambda-Support-in-debugger-expression-evaluator.md

```diff
@@ -0,0 +1,135 @@
+---
+title: "Lambda Support in debugger expression evaluator (Haruka Matsumoto)"
+---
+
+Author: Haruka Matsumoto
+
+# Abstract
+
+My project is lambda support in the MonoDevelop debugger's expression evaluator, used in the Immediate pad.
+
+# Code
+
+All of my work is covered by a single pull request: [here](https://github.com/mono/debugger-libs/pull/142#pull)
+
+# Summary of the implementation
+
+![Processing flow chart](/images/gsoc-2017-lambda-support-in-debugger expression-evaluator.png "Processing flow chart")
+
+*Fig 1. Processing flow chart*
+
+The debugger works using an interface [SDWP](http://www.mono-project.com/docs/advanced/runtime/docs/soft-debugger-wire-format/) exposed by Mono runtime. Using this protocol, the debugger controls debuggees and queries information of the debuggees like address of objects, type or methods. The debugger's evaluator usually works as an interpreter, which evaluates every part of expressions immediately. However, lambda support can not be realized in this way. We don't know when the debuggees invoke lambdas. The code of a lambda needs to stay invocable in the debuggees. This is how lambda expressions are different from other ones for the debugger's evaluator.
+
+There are roughly 2 tasks to support lambdas:
+
+* Task 1: Compile lambda expression using [Roslyn](https://github.com/dotnet/roslyn) API
+* Task 2: Inject compiled assembly to runtime through SDWP
+
+We firstly resolve references outside the lambda like local variables/properties (e.g., `x` in `a => a + x`) in order to reproduce the current context of the debuggee. Then we compile something like the following code with assemblies currently loaded in the debuggee, and get a .dll.
+
+```
+public class Injected_
+{
+	public static [lambda-type] injected_method([local-variables-list])
+	{
+		return [lambda-expression]
+	}
+}
+```
+
+`lambda-type` in the example above is a type name for lambda (e.g., `Func<int, int>`), `local-variables-list` is comma separated pairs of a type and variable name for local variables (e.g., `int y, int z`), and `lambda-expression` is a lambda inputted by users. (e.g., `x => x + y + z`)
+
+We send the compiled assembly to runtime, and invoke `Assembly.Load` with it in runtime. By calling `injected_method` through reflection APIs, the lambda value becomes accessible from the debugger.
+
+# Difficulties
+
+### Lambda types depends on debuggee's context
+
+Lambda type resolution is delayed in this project as it's not decidable without how it is used.
+When users input `(x => x + 5)`, it is undecidable which type is the most plausible out of following types.
+
+* System.Func<int, int>
+* System.Func<string, string>
+* System.Func<short, int>
```

Comment:
these should be wrapped in backticks as well.

Reply:
Thanks!